File: other/Solution221.java
import java.util.Stack;

class Solution221 {
    public int maximalSquare(char[][] matrix) {
    	int m = matrix.length;
    	int n = matrix[0].length;
    	int[][] heights = new int[m][n];
    	for(int i = 0; i < m; i++){
    		for(int j = 0; j < n; j++){
    			if(i == 0){
    				heights[i][j] = matrix[i][j] - '0';
    			}
    			else{
    				heights[i][j] = matrix[i][j] == '0' ? 0 : heights[i - 1][j] + 1;
    			}
    		}
    	}
    	int max = 0;
    	int[] left = new int[n];
    	int[] right = new int[n];
    	for(int i = 0; i < m; i++){
    		Stack<Integer> stack1 = new Stack<>();
    		Stack<Integer> stack2 = new Stack<>();
    		for(int j = 0; j < n; j++){
    			while(!stack1.empty() && heights[i][j] <= heights[i][stack1.peek()]){
    				stack1.pop();
    			}
    			left[j] = stack1.empty() ? -1 : stack1.peek();
    			stack1.push(j);
    		}

    		for(int j = n - 1; j >= 0; j--){
    			while(!stack2.empty() && heights[i][j] <= heights[i][stack2.peek()]){
    				stack2.pop();
    			}
    			right[j] = stack2.empty() ? n : stack2.peek();
    			stack2.push(j);
    		}
    		for(int j = 0; j < n; j++){
    			int length = right[j] - left[j] - 1;
    			if(length >= heights[i][j]){
    				max = Math.max(max, heights[i][j] * heights[i][j]);
    			}
    		}
    	}
    	return max;
    }

    public static void  main(String[] args){
    	Solution221 s = new Solution221();
    	char[][] test = {{'1', '0', '1', '0', '0'}, {'1', '0', '1', '1', '1'}, {'1', '1', '1', '1', '1'}, {'1', '0', '0', '1', '0'}};
    	System.out.println(s.maximalSquare(test));
    }
}

File: C++/Solution221.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<unordered_map>
#include<unordered_set>
#include<queue>
#include<stack>
using namespace std;

typedef long long ll;
typedef pair<int, int> pii;

class Solution221 {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int n = matrix.size(), m = matrix[0].size();
        vector<vector<int>> num(n, vector<int>(m));
        for(int j = 0; j < m; j++){
            num[0][j] = matrix[0][j] - '0';
        }
        for(int row = 1; row < n; row++){
            for(int col = 0; col < m; col++){
                if(matrix[row][col] == '0'){
                    num[row][col] = 0;
                }
                else{
                    num[row][col] = num[row - 1][col] + 1;
                }
            }
        }
        int ans = 0;
        for(int row = 0; row < n; row++){
            stack<int> stk;
            vector<int> left(m), right(m);
            for(int col = 0; col < m; col++){
                while(!stk.empty() && num[row][col] <= num[row][stk.top()]){
                    stk.pop();
                }
                if(stk.empty()){
                    left[col] = -1;
                }
                else{
                    left[col] = stk.top();
                }
                stk.push(col);
            }
            stack<int> stk2;
            for(int col = m - 1; col >= 0; col--){
                while(!stk2.empty() && num[row][col] <= num[row][stk2.top()]){
                    stk2.pop();
                }
                if(stk2.empty()){
                    right[col] = m;
                }
                else{
                    right[col] = stk2.top();
                }
                stk2.push(col);
            }
            for(int col = 0; col < m; col++){
                int x = min(num[row][col], right[col] - left[col] - 1);
                ans = max(ans, x * x);
            }
        }
        return ans;
    }
};
